src/java/org/apache/cassandra/cql3/transactions/RowDataReference.java

```diff
@@ -250,15 +251,15 @@ private void resolveFinished()
             isResolved = true;
         }
 
-        public void resolveReference(Map<TxnDataName, ReferenceSource> sources)
+        public void resolveReference(Map<Integer, ReferenceSource> sources, Map<String, Integer> nameToTxnDataName, int userReadIndex)
         {
             if (isResolved)
                 return;
 
-            // root level name
-            tupleName = TxnDataName.user(tuple.toString());
+            // root level name, use the one that was already created if it exists otherwise generate a new one
+            tupleName = nameToTxnDataName.getOrDefault(tuple.toString(), userReadIndex);
             ReferenceSource source = sources.get(tupleName);
-            checkNotNull(source, CANNOT_FIND_TUPLE_MESSAGE, tupleName.name());
+            checkNotNull(source, CANNOT_FIND_TUPLE_MESSAGE, tuple.toString());
```

Comment:
nit: maybe introduce an accessor that makes clear we're just extracting the existing prepared `text` field?